### PR TITLE
Try to fix https://github.com/csarofeen/pytorch/issues/286

### LINKED ISSUE #286: codegen error on broadcast
## 🐛 Bug

Simple c++ test fails to run. I have pushed the repro to https://github.com/csarofeen/pytorch/tree/bcast_codegen_error_repro

The error was introduced in our recent index fixing PR: 2dec1a5d065a3505df6969bd9a6bc80b5c4e579d
To be fair, prior to that PR, the generated code produces wrong results.

```
void testGPU_FusionComplexBCastFailed() {                                                                                                                                                                                                                                         
  Fusion fusion;                                                                                                            
  FusionGuard fg(&fusion);                                                                                                  
                                                                                                                            
  int w = 4, x = 8, y = 6, z = 8;                                                                                           
  DataType dtype = DataType::Float;                                                                                         
  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);                                                
                                                                                                                            
  at::Tensor t0 = at::randn({x, y, z}, options);                                                                            
  at::Tensor t1 = at::randn({w, x, y, z}, options);                                                                         
                                                                                                                            
  auto tv0 = makeDummyTensor(3);                                                                                            
  auto tv1 = makeDummyTensor(4);                                                                                            
  fusion.addInput(tv0);                                                                                                     
  fusion.addInput(tv1);                                                                                                     
                                                                                                                            
  auto tv2 = add(tv0, new Float(1.0));                                                                                      
  auto tv3 = add(tv2, tv1);                                                                                                 
                                                                                                                            
  fusion.addOutput(tv3);                                                                                                    
                                                                                                                            
  fuser::cuda::scheduleFusion(&fusion, {t0, t1});                                                                           
                                                                                                                            
  torch::jit::fuser::cuda::FusionExecutor fe;                                                                               
  fe.compileFusion(&fusion);                                                                                                
  auto outputs = fe.runFusion({t0, t1});                                                                                    
                                                                                                                            
  auto t2 = t0.add(1.0);                                                                                                    
  auto t3 = t2.add(t1);                                                                                                     
                                                                                                                            
  TORCH_CHECK(t3.allclose(outputs[0]));                                                                                     
}
```

It errors out with 
```
unknown file: Failure
C++ exception with description "index_map.find(kir_root_dom_i) != index_map.end() INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_compute.cpp":677, please report a bug to PyTorch. Couldn't find root mapping for TV2 dim: 0 id: iS{T4.size[1]}
Exception raised from getGlobalProducerIndex at ../torch/csrc/jit/codegen/cuda/index_compute.cpp:677 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f87a0e4e99b in /volume/codegen_project/pytorch_codegen/build/lib/libc10.so)
frame #1: <unknown function> + 0x2cfea43 (0x7f87a3d71a43 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #2: <unknown function> + 0x2d000e5 (0x7f87a3d730e5 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #3: torch::jit::fuser::IndexLowering::lowerOperand(torch::jit::fuser::Val*, torch::jit::fuser::Val*) const + 0x61 (0x7f87a3dc7c01 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #4: torch::jit::fuser::IndexLowering::handle(torch::jit::fuser::BinaryOp*) + 0x30 (0x7f87a3dc8460 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #5: void torch::jit::fuser::Expr::dispatch<torch::jit::fuser::OptInDispatch*>(torch::jit::fuser::OptInDispatch*, torch::jit::fuser::Expr*) + 0xc0 (0x7f87a3d22c40 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #6: torch::jit::fuser::IndexLowering::handle(torch::jit::fuser::kir::ForLoop*) + 0xc0 (0x7f87a3dc8340 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #7: void torch::jit::fuser::Expr::dispatch<torch::jit::fuser::OptInDispatch*>(torch::jit::fuser::OptInDispatch*, torch::jit::fuser::Expr*) + 0x168 (0x7f87a3d22ce8 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #8: torch::jit::fuser::IndexLowering::handle(torch::jit::fuser::kir::ForLoop*) + 0xc0 (0x7f87a3dc8340 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #9: void torch::jit::fuser::Expr::dispatch<torch::jit::fuser::OptInDispatch*>(torch::jit::fuser::OptInDispatch*, torch::jit::fuser::Expr*) + 0x168 (0x7f87a3d22ce8 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #10: torch::jit::fuser::IndexLowering::handle(torch::jit::fuser::kir::IfThenElse*) + 0xe8 (0x7f87a3dc81d8 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #11: void torch::jit::fuser::Expr::dispatch<torch::jit::fuser::OptInDispatch*>(torch::jit::fuser::OptInDispatch*, torch::jit::fuser::Expr*) + 0x180 (0x7f87a3d22d00 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #12: torch::jit::fuser::IndexLowering::handle(torch::jit::fuser::kir::ForLoop*) + 0xc0 (0x7f87a3dc8340 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #13: void torch::jit::fuser::Expr::dispatch<torch::jit::fuser::OptInDispatch*>(torch::jit::fuser::OptInDispatch*, torch::jit::fuser::Expr*) + 0x168 (0x7f87a3d22ce8 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #14: torch::jit::fuser::IndexLowering::generate(std::vector<torch::jit::fuser::Expr*, std::allocator<torch::jit::fuser::Expr*> > const&) + 0x27 (0x7f87a3dc8017 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #15: torch::jit::fuser::GPULower::lower() + 0x3a3 (0x7f87a3df28b3 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #16: torch::jit::fuser::cuda::FusionExecutor::compileFusion(torch::jit::fuser::Fusion*, torch::jit::fuser::cuda::CompileOptions) + 0x256 (0x7f87a3d37926 in /volume/codegen_project/pytorch_codegen/build/lib/libtorch_cuda.so)
frame #17: torch::jit::testGPU_FusionComplexBCastFailed() + 0x3ae (0x5561c516272e in ./test_jit)
frame #18: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x4a (0x5561c51ae0aa in ./test_jit)
frame #19: <unknown function> + 0x25c015 (0x5561c51a4015 in ./test_jit)
frame #20: <unknown function> + 0x25c625 (0x5561c51a4625 in ./test_jit)
frame #21: <unknown function> + 0x25c8d5 (0x5561c51a48d5 in ./test_jit)
frame #22: testing::internal::UnitTestImpl::RunAllTests() + 0xc1c (0x5561c51a592c in ./test_jit)
frame #23: testing::UnitTest::Run() + 0x98 (0x5561c51a5be8 in ./test_jit)
frame #24: main + 0xc8 (0x5561c4fc5fd8 in ./test_jit)
frame #25: __libc_start_main + 0xe7 (0x7f87a00f0b97 in /lib/x86_64-linux-gnu/libc.so.6)
frame #26: _start + 0x2a (0x5561c4fd388a in ./test_jit)
" thrown in the test body.

```

### PR DESCRIPTION
**Pull request by csarofeen**
Attempt to fix: https://github.com/csarofeen/pytorch/issues/286

### CODE REVIEW

**Comment by csarofeen** [comment]
CC @naoyam to remind me to talk this crazy PR through with him Monday.

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -5185,6 +5328,7 @@ void testGPU_FusionSmemBlockGemm() {
 }
 
 void testGPU_FusionSmemBlockGemmCache() {
+#if 0
```
Comment (jjsjann123) [inline test/cpp/jit/test_gpu.cpp:5331]: Should we open an issue to track the breakage of this test, just so we won't forgot that something is disabled in cpp tests via macro and test is giving us a pass.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 2]
```diff
@@ -79,7 +79,10 @@ class ContigIDs : public OptInDispatch {
     if (!(std::all_of(
             ordered_inputs.begin(),
             ordered_inputs.end(),
-            [this](IterDomain* id) { return is_contig_root.at(id); }))) {
+            [this](IterDomain* id) {
+              return is_contig_root.at(id) && !id->isBroadcast() &&
+                  !id->isReduction();
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:84]: Does this actually fix the issue with broadcasting & contiguous simultaneously marked on a tensor?
The issue I reported in #230 actually went away in this PR.

**Review by jjsjann123 — APPROVED** [review]
don't see anything wrong in this PR. Let's clean it up and get everything upstream!

**Pull request merged**